<div class="alert alert-info">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Привет. Меня зовут Алина Гусева, я буду делать ревью к твоей работе по исследовательскому анализу данных. В дальнейшем предлагаю общаться на "ты".
    
    
Видно, что ты старалась, делая проект. Все ключевые этапы проекта выполнены, в целом тебе удалось со всем справиться. Особенно хочется отметить, что соблюдена структура проекта, а код написан аккуратно.
    
Для того, чтобы и тебе и мне удобнее бы
    ло ориентироваться в комментариях, я буду помечать их разными цветами:
- красный цвет означает комментарии, которые требуют доработки. Как только ты их исправишь - проект будет принят;
- желтый цвет - то, что в следующий раз можно будет сделать по-другому. На твое усмотрение можешь доделать маленькие правки сейчас или учесть замечание для выполнения следующих проектов;
- зеленый цвет - очень элегантные и удачные решение, которые желательно использовать в дальнейшей работе.
- синий цвет - информация
    
Предлагаю работать над проектом в диалоге: если ты что-то изменяешь или вносишь какие-то коррективы - выделяй это каким-нибудь цветом и пиши об этом мне. Так мне гораздо легче будет отследить сделанные изменения.
    
Также попрошу тебя не изменять и не удалять мои комментарии, чтобы проверка твоего самостоятельного проекта прошла оперативнее.
    
 
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Описание данных:
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [66]:
#Прочитаем исходный файл

import pandas as pd
data = pd.read_csv('/datasets/data.csv')


In [67]:
#Проверим общую информацию о данных

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [68]:
#Переименуем столбец 'dob_years' в 'full_years'

data = data.rename(columns={'dob_years': 'full_years'})

In [69]:
#Посмотрим на первые 5 строк таблицы

display(data.head())

children  days_employed  full_years education  education_id  \
0         1   -8437.673028          42    высшее             0   
1         1   -4024.803754          36   среднее             1   
2         0   -5623.422610          33   Среднее             1   
3         3   -4124.747207          32   среднее             1   
4         0  340266.072047          53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Молодец, хороший тон использовать метод info() для получения общей информации из таблицы. Для более детального изучения можно воспользоваться методом describe(). 
    
 
</div>

In [70]:
#Изучим данные с помошью метода describe()

data.describe()

children  days_employed    full_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

<font color='purple'>Комментарий И.: Метод describe() применен для набора данных</font>


In [71]:
#Посчитаем сколько отрицательных значений в колонке days_employed

print(data[data['days_employed'] < 0]['income_type'].count())

15906


In [72]:
#Посчитаем минимальные и средние по данным трудового стажа по категориям

display(data.groupby('income_type').agg({'days_employed' : ['min', 'mean']}))


days_employed               
                           min           mean
income_type                                  
безработный      337524.466835  366413.652744
в декрете         -3296.759962   -3296.759962
госслужащий      -15193.032201   -3399.896902
компаньон        -17615.563266   -2111.524398
пенсионер        328728.720605  365003.491245
предприниматель    -520.848083    -520.848083
сотрудник        -18388.949901   -2326.499216
студент            -578.751554    -578.751554

**Вывод**

На основании общей информации, можно сделать вывод, что в столбцах days_employed и total_income есть пропущенные значения (по 19351 записей, в то время как в других столбцах 21525 записей). Пропущенные значения необходимо удалить или заполнить.

Ошибок в названиях столбцов не обнаружено, они передают смысл содержания (есть сомнения по поводу столбца dob_years, кто такой "dob"? почему не "full"? Решено переименовать). По необходимости в названиях используется знак нижнего подчеркивания.

На данном этапе кажется необычным, что тип данных в столбце total_income и days_employed - float64, необходимо понять логику расчета и изменить. 

В столбце days_employed встречаются отрицательные значения, что видно на примере первых 5 строк, это ошибочные данные их необходимо изменить. Поскольку в колонке days_employed 15906 строк с отрицательными данными, мы не можем их просто удалить. Есть предположение, что количество стажа исчисляется как разница между датой трудоустройства и датой выгрузки данных, необходимо уточнить этот вопрос с разработчиками. По категориям 'пенсионер' и 'безработный' трудовой стаж исчисляется очень большими числами. 

<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Действительно, ты верно заметила, что количество проусков в двух столбцах одинаковое. Осталось проверить, взаимосвязаны ли они, то есть пропуски в одном соответсвуют пропускам в другом столбце.
    
    
Согласна с тобой, что есть столбцы, которые названы не корректно, можно заменять на более понятные названия.
    

Тип данных необходимо изменить, сейчас формат float из-за того, что в этих столбцах находятся значения NaN.
    
 
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

In [73]:
#Получим первые 10 записей с пропущенными значениями

display(data[data['days_employed'].isnull()].head(10))


children  days_employed  full_years education  education_id  \
12         0            NaN          65   среднее             1   
26         0            NaN          41   среднее             1   
29         0            NaN          63   среднее             1   
41         0            NaN          50   среднее             1   
55         0            NaN          54   среднее             1   
65         0            NaN          21   среднее             1   
67         0            NaN          52    высшее             0   
72         1            NaN          32    высшее             0   
82         2            NaN          50    высшее             0   
83         0            NaN          52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

In [74]:
#Заменим отрицательные значения в столбце days_employed на положительные

#data.loc[data['days_employed'] < 0, 'days_employed'] = data['days_employed']*(-1)

data.loc[data['days_employed'] < 0, 'days_employed'] = data['days_employed'].abs()

<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Конечно, тут необходима замена на положительные значения. Ты все сделала правильно. но гораздо проще можно получить такой же результат, используя метод abs(), который возвращает модуль числа.
 
</div>

<font color='purple'>Комментарий И.: Спасибо, применен метод abs()</font>


In [75]:
#Проверим результат путем вызова средних по категориям

print(data.groupby('income_type')['days_employed'].mean())

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          3399.896902
компаньон            2111.524398
пенсионер          365003.491245
предприниматель       520.848083
сотрудник            2326.499216
студент               578.751554
Name: days_employed, dtype: float64


In [76]:
#Проверим сколько записей пропущено в столбцах 'days_employed' и 'total_income'

print(len(data[data['days_employed'].isnull()]))
print(len(data[data['total_income'].isnull()]))

2174
2174


<div class="alert alert-warning">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Хм... а почему ты заменила стаж этих категорий на 0? Получается что пенсионер совсем не работал? Да, на данный момент эта категория на пенсии, но в этой колонке проставляется суммарный стаж проработанных лет. В данном случае такую замену делать некорректно. Подумай, как еще можно заменить пропущенный стаж?
    
 
</div>

<font color='purple'>Комментарий И.: Наверное можно заменить пропущенный стаж для категорий 'пенсионер' и 'безработный' на средний стаж по всем остальным категориям (обнуление стажа для пенсионера и безработного оставлено первым шагом, чтобы далее посчитать среднее по остальным данным, также можно было бы при подсчете средного добавить условия income_type 'не равно' пенсионер и безработный, чтобы среднее по остальным категориям посчиталось корректно)</font>


In [77]:
#Заменим стаж пенсионеров и безработных на средний стаж работы:
days_employed_avg = data['days_employed'].mean()

data.loc[(data['income_type'] == 'пенсионер'), 'days_employed'] = days_employed_avg
data.loc[(data['income_type'] == 'безработный'), 'days_employed'] = days_employed_avg


In [78]:
#Проверим, что данные обновились

print(data[data['income_type'] == 'пенсионер']['days_employed'].mean())
print(data[data['income_type'] == 'безработный']['days_employed'].mean())

66914.72890682239
66914.72890682236


<div class="alert alert-warning">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Код будет выглядеть проще, если ты убререшь groupby('income_type'). Ведь с маской ['income_type'] == 'безработный'] ты и так отбираешь только одну группу, соотвсетвенно группировка тут явно лишняя.
    
 
</div>

<font color='purple'>Комментарий И.: Исправлено</font>


In [79]:
#Посчитаем сколько записей в каждой категории

print(data['income_type'].value_counts())

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64


In [80]:
#Объединим категории с небольшим количеством записей

data.loc[data['income_type'] == 'предприниматель', 'income_type'] = 'компаньон'
data.loc[data['income_type'] == 'безработный', 'income_type'] = 'другое'
data.loc[data['income_type'] == 'в декрете', 'income_type'] = 'другое'
data.loc[data['income_type'] == 'студент', 'income_type'] = 'другое'

In [81]:
#Проверим новый список категорий

print(data['income_type'].value_counts())

сотрудник      11119
компаньон       5087
пенсионер       3856
госслужащий     1459
другое             4
Name: income_type, dtype: int64


In [82]:
#Посчитаем медиану и среднее для столбца 'total_income' в зависимости от категории

print(data.groupby('income_type').agg({'total_income': ['mean','median']}))


              total_income               
                      mean         median
income_type                              
госслужащий  170898.309923  150447.935283
другое       103677.564849   79079.308649
компаньон    202482.281401  172396.000846
пенсионер    137127.465690  118514.486412
сотрудник    161380.260488  142594.396847


In [83]:
#Посчитаем медиану и среднее для столбца 'days_employed' в зависимости от категории

print(data.groupby('income_type').agg({'days_employed': ['mean','median']}))

            days_employed              
                     mean        median
income_type                            
госслужащий   3399.896902   2689.368353
другое       34426.242332  35105.744434
компаньон     2111.176937   1546.333214
пенсионер    66914.728907  66914.728907
сотрудник     2326.499216   1574.202821


In [84]:
#Проверим для каких категорий есть пропущенные значения в столбце 'total_income'

print(data[data['total_income'].isnull()].groupby('income_type')['income_type'].count())

income_type
госслужащий     147
компаньон       509
пенсионер       413
сотрудник      1105
Name: income_type, dtype: int64


In [85]:
#Проверим для каких категорий есть пропущенные значения в столбце 'days_employed'

print(data[data['days_employed'].isnull()].groupby('income_type')['income_type'].count())

income_type
госслужащий     147
компаньон       509
сотрудник      1105
Name: income_type, dtype: int64


In [86]:
#Заменим пропущенные значения в столбце 'total_income' для категорий госслужащий, компаньон, пенсионер, сотрудник на медианные значения по каждой категории
#median_income_gov = data[data['income_type'] == 'госслужащий']['total_income'].median()
#median_income_comp = data[data['income_type'] == 'компаньон']['total_income'].median()
#median_income_ret = data[data['income_type'] == 'пенсионер']['total_income'].median()
#median_income_emp = data[data['income_type'] == 'сотрудник']['total_income'].median()

#data.loc[(data['income_type'] == 'госслужащий') & (data['total_income'].isna()), 'total_income'] = median_income_gov
#data.loc[(data['income_type'] == 'компаньон') & (data['total_income'].isna()), 'total_income'] = median_income_comp
#data.loc[(data['income_type'] == 'пенсионер') & (data['total_income'].isna()), 'total_income'] = median_income_ret
#data.loc[(data['income_type'] == 'сотрудник') & (data['total_income'].isna()), 'total_income'] = median_income_emp

med = data.groupby('income_type')['total_income'].transform('median')
data['total_income'] = data['total_income'].fillna(med)


<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Отлично, все верно! Но опять же много кода. Здесь у тебя много повторяющихся строк с разным income_type, это должно натолкнуть на мысль, что можно написать свою функцию, которая будет возвращать нужный результат. Онако в питоне уже есть стандартная функция <a href='https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.transform.html'>transform()</a>, которая тут будет как раз к месту.
</div>

<font color='purple'>Комментарий И.: Попыталась применить функцию transform. Приятно видеть две строчки кода вместо 8!</font>


<div class="alert alert-success">
<h2> Комментарий от ревьюера 2<a class="tocSkip"></a></h2>
Молодец, теперь код вылядит гораздо компактнее)
</div>

In [87]:
#Проверим наличие пропусков в столбце 'total_income'

data['total_income'].isna().sum()

0

In [88]:
#Заменим пропущенные значения в столбце 'days_employed' для категорий госслужащий, компаньон, сотрудник на медианные значения по каждой категории
#median_days_gov = data[data['income_type'] == 'госслужащий']['days_employed'].median()
#median_days_comp = data[data['income_type'] == 'компаньон']['days_employed'].median()
#median_days_emp = data[data['income_type'] == 'сотрудник']['days_employed'].median()

#data.loc[(data['income_type'] == 'госслужащий') & (data['days_employed'].isna()), 'days_employed'] = median_days_gov
#data.loc[(data['income_type'] == 'компаньон') & (data['days_employed'].isna()), 'days_employed'] = median_days_comp
#data.loc[(data['income_type'] == 'сотрудник') & (data['days_employed'].isna()), 'days_employed'] = median_days_emp

med = data.groupby('income_type')['days_employed'].transform('median')
data['days_employed'] = data['days_employed'].fillna(med)


In [89]:
#Проверим наличие пропусков в data

data.isna().sum()

children            0
days_employed       0
full_years          0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [90]:
#Проверим уникальные значения в столбце долга

data['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

In [91]:
#Проверим уникальные значения в столбце с количеством детей

data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [92]:
data['children'] = data['children'].replace(to_replace=[-1,20],value=[1,2])
data['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Таких строк мало, можно и удалять, код будет выглядеть более читаемый, если ты воспользкешься функцией drop(). Но лучше все-таки сделать замену -1 на 1 и 20 на 2.
</div>

<font color='purple'>Комментарий И.: Данные заменены на 1 и 2</font>


<div class="alert alert-success">
<h2> Комментарий от ревьюера 2<a class="tocSkip"></a></h2>
Кстати, вот тут тоже можно использовать python-way, воспользоваться функцией replace() для замены необходимых значений.
</div>

**Вывод**

Для выявления пропущенных значений использовался метод isnull().

На примере первых десяти строк таблицы с пропущенными значениями можно предположить, что значения одновременно пропущены в столбцах days_employed и total_income. 
Посчитаем количество пропусков в этих двух столбцах.
Маловероятно, что люди не хотели делиться данными о своем доходе при заполнении заявки на кредит. Данные могли быть внесены и/или выгружены некорректно, причину пропуска в данных можно проверить дополнительно с разработчиками. Пока к разработчикам доступа нет, необходимо заменить данные.

Поскольку пропущены количественные данные, принято решение посчитать средний и медианных доход и количество дней трудоустройства в зависимости от типа занятости и произвести замену пропущенных значений.
Для того, чтобы правильно посчитать среднее по количеству дней трудоустройства, сначала обработаем отрицательные значения и большие числа в этом столбце.
Для пенсионера и безработного заменим количество days_employed на 0. Отрицательные значения переведем в положительные - предположительно из-за логики сбора информации в этом столбце значения необходимо брать по модулю.


Проверим какие есть категории и сколько человек в каждой категории. Категории 'безработный', 'предприниматель', 'в декрете', 'студент' значительно меньше первых четырех категорий. Объединим категории 'безработный', 'в декрете', 'студент' в категорию 'другое'. Предпринимателя поместим в категорию 'компаньон' из-за высокого дохода.       
Сравним средние и медианные значения по категориям. Видно, что средние значения смещены вправо относительно медианных, вероятно в каждой категории есть зарплаты и длительность работы значительно выше остальных. 
Для замены пропущенных значений принято решение использовать медианные значения.

Проверка столбца долга показала, что в столбце есть только два значения, что свидетельствует о том, что данные правильные.
В столбце с наличием детей в 47 строках указано значение '-1' и в 76 строках указано значение '20'. Принято решение удалить эти строки, поскольку отрицательные данные ошибочные, и данные 20 вероятно тоже (такое возможно, но вряд ли в 76 случаях).




### Замена типа данных

In [93]:
#Заменим тип данных в столбце 'total_income' на целочисленный

data['total_income'] = data['total_income'].astype('int')

In [94]:
#Заменим тип данных в столбце 'days_employed' на целочисленный

data['days_employed'] = data['days_employed'].astype('int')

In [95]:
#Проверим общую инфомрацию по таблице

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   full_years        21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

Для столбцов 'total_income' и 'days_employed' произведена замена типа данных с вещественнного на целочисленный с помощью метода astype().





<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Отлично,теперь все столбцы имеют корректный формат данных.
</div>

### Обработка дубликатов

In [96]:
#Проверим сколько в данных содержится дубликтов

data.duplicated().sum()

54

In [97]:
#Проверим как выглядят категории в столбце 'education'

data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

In [98]:
#Переведем строки в столбце data в нижний регистр

data['education'] = data['education'].str.lower()


In [99]:
#Проверим сколько в данных содержится дубликатов после изменения регистра столбца education

data.duplicated().sum()

71

<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Здесь важно обратить внимание на то, что сначала необходимо привести текстовый формат данных к единому виду(регистр), а затем проверять на дубликаты. Иначе результат может получится другой.
 
</div>

In [100]:
#Проверим как выглядят категории в столбце 'family_status'

data['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [101]:
#Проверим как выглядят категории в столбце 'purpose'

data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

In [102]:
#Удалим дубликаты из данных

data = data.drop_duplicates().reset_index(drop=True)

In [103]:
#Проверим наличие дубликатов после удаления

data.duplicated().sum()

0

**Вывод**

Был обнаружен и удален 71 дубликат.
Возможно, дубликаты появились в следствие повторного обращения людьми за кредитом.
Помимо этого обнаружено, что некоторые цели кредита сформулированы по-разному, но имеют одинаковый смысл.


### Лемматизация

In [104]:
#Импортируем библиотеку для лемматизации 

from pymystem3 import Mystem
m = Mystem() 

In [105]:
#Лемматизируем значения в столбце с целями кредита
text = data['purpose'].unique()
text = ' '.join(text)

lemmas = m.lemmatize(text)
print(lemmas) 

                

['покупка', ' ', 'жилье', ' ', 'приобретение', ' ', 'автомобиль', ' ', 'дополнительный', ' ', 'образование', ' ', 'сыграть', ' ', 'свадьба', ' ', 'операция', ' ', 'с', ' ', 'жилье', ' ', 'образование', ' ', 'на', ' ', 'проведение', ' ', 'свадьба', ' ', 'покупка', ' ', 'жилье', ' ', 'для', ' ', 'семья', ' ', 'покупка', ' ', 'недвижимость', ' ', 'покупка', ' ', 'коммерческий', ' ', 'недвижимость', ' ', 'покупка', ' ', 'жилой', ' ', 'недвижимость', ' ', 'строительство', ' ', 'собственный', ' ', 'недвижимость', ' ', 'недвижимость', ' ', 'строительство', ' ', 'недвижимость', ' ', 'на', ' ', 'покупка', ' ', 'подержать', ' ', 'автомобиль', ' ', 'на', ' ', 'покупка', ' ', 'свой', ' ', 'автомобиль', ' ', 'операция', ' ', 'с', ' ', 'коммерческий', ' ', 'недвижимость', ' ', 'строительство', ' ', 'жилой', ' ', 'недвижимость', ' ', 'жилье', ' ', 'операция', ' ', 'со', ' ', 'свой', ' ', 'недвижимость', ' ', 'автомобиль', ' ', 'заниматься', ' ', 'образование', ' ', 'сделка', ' ', 'с', ' ', 'подержать

In [106]:
#Посчитаем сколько раз каждое слово употреблялось в столбце purpose

from collections import Counter
print(Counter(lemmas))

Counter({' ': 96, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'подержать': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'со': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1, '\n': 1})


**Вывод**

Для лемматизации целей кредита была загружена библиотека pymystem3,был сформирован список уникальных значений и переведен в строку с помощью метода join(). Затем к тексту в строке был применен метод lemmatize().
С помощью Counter были посчитаны повторяющиеся слова.
В итоге можно выделить следующие группы целей кредита:
недвижимость (включая жилье), автомобиль, образование, свадьба.


<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Ура, лемматизация прошла успешно! Удалось выделить основные причины получения кредита.
 
</div>

### Категоризация данных

In [107]:
#Сформируем словарь для семейного положения

#family_status_dict = data[['family_status', 'family_status_id']]
#family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
#print(family_status_dict)

<div class="alert alert-warning">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Можно это действие опустить, так как в изначальной таблице словат=рь с семейный статусом уже был.
 
</div>

<font color='purple'>Комментарий И.: Согласна, действие скрыто</font>


In [108]:
#Для столбца с целями кредита напишем формулу, которая добавляет категорию кредита для каждой цели

def purposes_dict (purpose):
    """
    Возвращает категорию цели покупки в зависимости от определяющего слова в исходной цели.
    Если исходная цель содержит часть слова 'автомобил' - присвается категория 'автомобиль'
    Если исходная цель содержит часть слова 'свадьб' - присвается категория 'свадьба'
    Если исходная цель содержит часть слова 'образован' - присвается категория 'образование'
    Если исходная цель содержит часть слова 'недвижимост' - присвается категория 'недвижимость'
    Если исходная цель содержит часть слова 'жиль' - присвается категория 'недвижимость'
    """
    
    if 'автомобил' in purpose:
        return 'автомобиль'
    if 'свадьб' in purpose:
        return 'свадьба'
    if 'образован' in purpose:
        return 'образование'
    if 'недвижимост' or 'жиль' in purpose:
        return 'недвижимость'
    
    
data['purpose_group'] = data['purpose'].apply(purposes_dict)
display(data.head(10))

children  days_employed  full_years education  education_id  \
0         1           8437          42    высшее             0   
1         1           4024          36   среднее             1   
2         0           5623          33   среднее             1   
3         3           4124          32   среднее             1   
4         0          66914          53   среднее             1   
5         0            926          27    высшее             0   
6         0           2879          43    высшее             0   
7         0            152          50   среднее             1   
8         2           6929          35    высшее             0   
9         0           2188          41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_group  
0               покупка жилья  недвижимость  
1     приобретение автомобиля    автомобиль  
2               покупка жилья  недвижимость  
3  дополнительное образование   образование  
4             сыграть свадьбу       свадьба  
5               покупка жилья  недвижимость  
6           операции с жильем  недвижимость  
7                 образование   образование  
8       на проведение свадьбы       свадьба  
9     покупка жилья для семьи  недвижимость

In [109]:
#Проверим, что каждой цели присвоена группа

print(data['purpose_group'].isna().sum())

0


In [110]:
#Проверим разброс общего дохода

data_min = data['total_income'].min()
data_quantile_lower = data['total_income'].quantile(0.33)
data_median  = data['total_income'].median()
data_quantile_upper = data['total_income'].quantile(0.66)
data_max = data['total_income'].max()


print(data_min)
print(data_quantile_lower)
print(data_median)
print(data_quantile_upper)
print(data_max) 

20667
118514.0
142594.0
172396.0
2265604


<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
А вот это очень хороший способ, молодец! Теперь у тебя клиенты разделены на равные группы по уровню дохода.
    
Только дальнше в функции лучше не руками вбивать числа, а зарманее завести переменную. Допустим, что датафрейм немного изменится и тогда показатели уже будут другие.
</div>

<font color='purple'>Комментарии И.: Вместо значений в формулу подставлены переменные</font>


In [111]:
#Напишем формулу для категоризации дохода

def income_dict (total_income):
    """"
    Присваивает категорию дохода в зависимости от его размера
    Если доход менее значения первой квантили - присваивается значение 'низкий'
    Если доход от первой квантили включительно до второй квантили - присваивается значение 'средний'
    Если доход от второй квантили включительно - присваивается значение 'высокий'
    """
    if total_income < data_quantile_lower:
        return 'низкий'
    if data_quantile_lower <= total_income < data_quantile_upper:
        return 'средний'
    if total_income >= data_quantile_upper:
        return 'высокий'
    
data['income_group'] = data['total_income'].apply(income_dict)
display(data.head(10))    

children  days_employed  full_years education  education_id  \
0         1           8437          42    высшее             0   
1         1           4024          36   среднее             1   
2         0           5623          33   среднее             1   
3         3           4124          32   среднее             1   
4         0          66914          53   среднее             1   
5         0            926          27    высшее             0   
6         0           2879          43    высшее             0   
7         0            152          50   среднее             1   
8         2           6929          35    высшее             0   
9         0           2188          41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_group income_group  
0               покупка жилья  недвижимость      высокий  
1     приобретение автомобиля    автомобиль       низкий  
2               покупка жилья  недвижимость      средний  
3  дополнительное образование   образование      высокий  
4             сыграть свадьбу       свадьба      средний  
5               покупка жилья  недвижимость      высокий  
6           операции с жильем  недвижимость      высокий  
7                 образование   образование      средний  
8       на проведение свадьбы       свадьба       низкий  
9     покупка жилья для семьи  недвижимость      средний

In [48]:
#Проверим, что каждому доходу присвоена группа

print(data['income_group'].isna().sum())

0


**Вывод**

Для того, чтобы ответить на поставленные вопросы была проведена категоризация данных, а именно был создан словарь для семейного положения, были созданы и добавлены категории по целям кредита и общему доходу.
Для категоризации данных по целям кредита были выделены четыре основные группы: недвижимость (включая жилье), автомобиль, образование, свадьба, была создана формула, которая по части слова в исходных данных столбца целей относила цель к одной из 4х категорий.
Для категоризации дохода были посчитаны минимальный, медианный, максимальный доход, а также квартили 25% и 75%, затем была создана формула, с помощью которой каждой строке дохода присваивается категории в зависимости от размера дохода.


## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [49]:
#Построим сводную таблицу, показывающую данные о долгах в разрезе количества детей

data_pivot_children = data.pivot_table(index=['children'], columns = 'debt', values = 'income_type', aggfunc = 'count')

data_pivot_children['ratio'] = data_pivot_children[0] / (data_pivot_children[1] + data_pivot_children[0])
print(data_pivot_children.sort_values(by = 'ratio', ascending = False))

debt            0       1     ratio
children                           
0         13028.0  1063.0  0.924562
3           303.0    27.0  0.918182
1          4410.0   445.0  0.908342
2          1926.0   202.0  0.905075
4            37.0     4.0  0.902439
5             9.0     NaN       NaN


**Вывод**

Результаты анализа показывают, что люди без детей чаще всего берут кредит и при этом доля вернувших кредит у них самая высокая. Далее чаще всего возвращают кредит в срок люди с 3мя детьми, люди с 5ю детьми являются самой малочисленной группой заемщиков и при этом возвращают кредит всегда. У людей с 1, 2 и 4мя детьми примерно одинаковое соотношение возвратов и задолженностей.

<div class="alert alert-danger">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>
Логика правильная, но есть небольшая ошибка в формуле. Посмотри внимательно и подумай, ведь тебе нужно найти долю должиков, а в знаменателе у тебя стоит что-то не то. Это касается и остальных ответов на вопрос.
</div>

<div class="alert alert-success">
<h2> Комментарий от ревьюера 2<a class="tocSkip"></a></h2>
Да, заемщики без детей самые отвественные.
    

</div>

<font color='purple'>Комментарии И.: В знаменатель включена сумма должников и сумма выплативших кредит вовремя - общая сумма заемщиков по категории</font>


- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [56]:
#Построим сводную таблицу, показывающую данные о долгах в разрезе семейного статуса

data_pivot_status = data.pivot_table(index=['family_status'], columns = 'debt', values = 'income_type', aggfunc = 'count')

data_pivot_status['ratio'] = data_pivot_status[0] / (data_pivot_status[0] + data_pivot_status[1])
print(data_pivot_status.sort_values(by = 'ratio', ascending = False))

debt                       0    1     ratio
family_status                              
вдовец / вдова           896   63  0.934307
в разводе               1110   85  0.928870
женат / замужем        11408  931  0.924548
гражданский брак        3763  388  0.906529
Не женат / не замужем   2536  274  0.902491


**Вывод**

Данные свидетельствуют о том, что доля людей вернувших кредит выше всего у овдовевших людей и людей в разводе (возможно, этому как-то способствуют наследство и полученное имущество после его раздела). Наибольшее соотношение должников в группах людей, живущих в гражданском браке или не состоявших в браке. Меньше ответственности в личной жизни - меньше дисциплины в выплате кредита.

<div class="alert alert-success">
<h2> Комментарий от ревьюера 2 <a class="tocSkip"></a></h2>
Посмотрев на сводную таблицу можно увидеть, что доля просроченых кредитов больше у людей, которые находятся в гражданском браке или не женаты. Чаще всего возращают кредиты вдовцы и вдовы.
</div>


- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [57]:
#Построим сводную таблицу, показывающую данные о долгах в разрезе общего дохода

data_pivot_income = data.pivot_table(index=['income_group'], columns = 'debt', values = 'income_type', aggfunc = 'count')
data_pivot_income['ratio'] = data_pivot_income[0] / (data_pivot_income[1] + data_pivot_income[0])
print(data_pivot_income.sort_values(by = 'ratio', ascending = False))

debt             0    1     ratio
income_group                     
высокий       6958  558  0.925758
низкий        6165  544  0.918915
средний       6590  639  0.911606


**Вывод**

Предсказуемо в группе с высоким доходом отношение выплачивающих кредит вовремя самое высокое. Удивительно, что на втором месте идут люди с низким доходов. Видимо, низкие доходы приучают к финансовой дисциплине и планированию своих средства, что сказывается на своевременных выплатах. У людей со средним доходом соотношение самое низкое, с доходами растут потребности, и иногда они растут непропорционально быстро, что негативно влияет на выплаты по кредиту.

<div class="alert alert-success">
<h2> Комментарий от ревьюера 2<a class="tocSkip"></a></h2>
Молодец, верный вывод:)
</div>


- Как разные цели кредита влияют на его возврат в срок?

In [58]:
#Построим сводную таблицу, показывающую данные о долгах в разрезе целей кредита

data_pivot_purpose = data.pivot_table(index=['purpose_group'], columns = 'debt', values = 'income_type', aggfunc = 'count')
data_pivot_purpose['ratio'] = data_pivot_purpose[0] / (data_pivot_purpose[1] + data_pivot_purpose[0])
print(data_pivot_purpose.sort_values(by = 'ratio', ascending = False))

debt               0    1     ratio
purpose_group                      
недвижимость   10029  782  0.927666
свадьба         2138  186  0.919966
образование     3643  370  0.907800
автомобиль      3903  403  0.906410


**Вывод**

Люди, которые берут кредит под недвижимость являются самой большой группой и самыми организованными в отношении выплаты кредита. Затем, люди, берущие кредит на свадьбы (неужели еще такие есть). Ниже отношение у людей, берущих кредит на образование, скорее всего существенная часть группы это молодые люди с пока еще нестабильным доходом, что объясняет задержки в выплате кредита. Чаще всего задерживают с выплатами люди из группы заемщиков средств на автомобиль.

## Шаг 4. Общий вывод

Идеальный заемщик - это вдовец с 5ю детьми или без детей, с высоким доходом, покупающий недвижимость.
При выдаче кредита стоит перепроверить надежность человека, не состоящего в браке, у которого 4ро детей, средний доход и потребность в новом автомобиле.

<div class="alert alert-success">
<h2> Комментарий от ревьюера<a class="tocSkip"></a></h2>

Поздравляю с первым проектом!

- Ты хорошо делаешь выводы и анализируешь результаты.
    
- Код написан аккуратно, однако еще есть куда расти:)

- Соблюдена структура проекта, спасибо, было приятно проверять.
    
- Осталось исправить замечания и проект будет принят
 


Желаю дальнейших успехов!


</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.